modules/packages/BLAS.chpl

```diff
@@ -683,691 +718,1646 @@ module BLAS {
+      }
+      otherwise {
+        compilerError("Unknown type in syr: ", eltType:string);
+      }
+    }
+  }
+
+  /*
+    Wrapper for `SYR2`_ routines::
+
+      A := alpha*x*y'+ alpha*y*x' + A
+
+  */
+  proc syr2(A: [?Adom] ?eltType, X: [?vDom] eltType, Y: [vDom] eltType,
+            alpha,
+            order : Order = Order.Row,
+            uplo : Uplo = Uplo.Upper,
+            ldA : int = 0, incx : c_int = 1, incy : c_int = 1)
+            where (Adom.rank == 2) && (vDom.rank == 1)
+  {
+
+    var m = Adom.dim(1).size : c_int,
+        n = Adom.dim(2).size : c_int;
+
+    if m != n then
+      halt("Non-square array of dimensions %ix%i passed to syr2".format(m, n));
+
+    var _ldA = getLeadingDim(Adom, order, ldA);
+
+    select eltType {
+      when real(32) {
+        C_BLAS.cblas_ssyr2(order, uplo, n, alpha, X, incx, Y, incy, A, _ldA);
+      }
+      when real(64) {
+        C_BLAS.cblas_dsyr2(order, uplo, n, alpha, X, incx, Y, incy, A, _ldA);
+      }
+      otherwise {
+        compilerError("Unknown type in syr2: ", eltType:string);
+      }
+    }
+  }
+
+  pragma "no doc"
+  /*
+    Wrapper for the `TBMV`_ routines::
+
+      x := A*x,         when trans = Op.N
+      x := A'*x,        when trans = Op.T
+      x := conjg(A')*x, when trans = Op.H
+
+    TODO -- Get banded array routines working
+
+  */
+  proc tbmv(A: [?Adom] ?eltType, X: [?vDom] eltType,
+            k: c_int,
+            trans : Op = Op.N,
+            order : Order = Order.Row,
+            uplo : Uplo = Uplo.Upper,
+            diag : Diag = Diag.NonUnit,
+            ldA : int = 0, incx : c_int = 1)
+            where (Adom.rank == 2) && (vDom.rank == 1)
+  {
+    // Determine sizes
+    var m = Adom.dim(1).size : c_int,
+        n = Adom.dim(2).size : c_int;
+
+    // Set strides if necessary
+    var _ldA = getLeadingDim(Adom, order, ldA);
+
+    select eltType {
+      when real(32) {
+        C_BLAS.cblas_stbmv(order, uplo, trans, diag, m, k, A, _ldA, X, incx);
+      }
+      when real(64) {
+        C_BLAS.cblas_dtbmv(order, uplo, trans, diag, m, k, A, _ldA, X, incx);
+      }
+      when complex(64) {
+        C_BLAS.cblas_ctbmv(order, uplo, trans, Diag, m, k, A, _ldA, X, incx);
+      }
+      when complex(128) {
+        C_BLAS.cblas_ztbmv(order, uplo, trans, Diag, m, k, A, _ldA, X, incx);
+      }
+      otherwise {
+        compilerError("Unknown type in tbmv: ", eltType:string);
+      }
+    }
+  }
+
+  pragma "no doc"
+  /*
+   Wrapper for the `TBSV`_ routines::
+
+    A*x = b,         when trans = Op.N
+    A'*x = b,        when trans = Op.T
+    conjg(A')*x = b, when trans = Op.H
+
+   TODO -- Get banded array routines working
+
+  */
+  proc tbsv(A: [?Adom] ?eltType, X: [?vDom] eltType,
+            trans : Op = Op.N,
+            order : Order = Order.Row,
+            uplo : Uplo = Uplo.Upper,
+            diag : Diag = Diag.NonUnit,
+            ldA : int = 0, incx : c_int = 1)
+            where (Adom.rank == 2) && (vDom.rank == 1)
```

Comment:
I wonder if these 'where' clauses would be better as compilerError checks in the function body. In its current state someone would run into an unresolved call error, which isn't exactly all that useful.

What do you think?